@bobrusha Александра Боброва

bobrusha opened the pull request:
1) Пофиксила баг с обновлением битмапов несколько раз подряд. Когда делаешь скролл вниз, а потом скроллишь обратно, иногда сеттятся битмапы, отличные от показанных в первый раз, но из нужного жанра (я проверяла). Это связано с тем, что картинки загружаются каждый раз и параллельно, следовательно порядок битмапов в листе недетерменирован и может меняться. 
Описанное выше можно посмотреть [здесь](https://www.dropbox.com/s/189gghyazy7w2mz/demo.mp4?dl=0)

2) Критическая секция тоже работает. Пофиксила баг: когда было несколько скроллов, после того как скролл заканчивался, битмапы не сеттились

bobrusha committed:
Trying to understand what to do....

bobrusha commented:
![screen shot 2016-07-26 at 00 05 19](https://cloud.githubusercontent.com/assets/5622535/17117441/b8eae694-52c4-11e6-85ed-d65d24bf1ab5.png)

bobrusha committed:
Started 2nd task

bobrusha committed:
Changed to Rx

bobrusha committed:
Deleted threads

bobrusha committed:
Moved subscription to VH

bobrusha committed:
Minor improvements.

kotiknorm commented on app/src/main/java/ru/yandex/yamblz/loader/CollageLoader.java:
```diff
@@ -2,15 +2,18 @@
     void loadCollage(List<String> urls, ImageTarget imageTarget);
 
-    void loadCollage(List<String> urls, ImageView imageView, CollageStrategy collageStrategy);
+    Subscription loadCollage(List<String> urls, WeakReference<ImageView> imageView, CollageStrategy collageStrategy);
```
Можно же оборачивать в вик референс внутри метода, зачем тащить наружу детали реализации

bobrusha replied:
✅

kotiknorm commented on app/src/main/java/ru/yandex/yamblz/ui/adapters/GenreListAdapter.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Override
+    public void onBindViewHolder(GenreViewHolder holder, int position) {
+        if (holder.getSubscription() != null && holder.getSubscription().isUnsubscribed()) {
```
Условие точно должно быть "если не нулл и отписан"?

bobrusha replied:
упc
✅

kotiknorm commented on app/src/main/java/ru/yandex/yamblz/ui/adapters/GenreListAdapter.java:
```diff
@@ -0,0 +1,99 @@
+        holder.name.setText(genre.getName());
+        Log.d(DEBUG_TAG, Arrays.deepToString(genre.getUrls().toArray()));
+
+        Subscription subscription = CollageLoaderManager.getLoader()
```
Перед загрузкой нужно очищать ImageView  от старой картинки

kotiknorm replied:
В пикассо ты не знаешь о деталях реализации, а тут вынесла подписки и викреференсы, не очень гуд.

kotiknorm commented on app/src/main/java/ru/yandex/yamblz/loader/StubCollageLoader.java:
```diff
@@ -17,9 +33,35 @@ public void loadCollage(List<String> urls, ImageTarget imageTarget) {
+                    public void onCompleted() {
+                        Log.d(DEBUG_TAG, "Bitmap size + " + bitmaps.size());
+                        Bitmap collage = collageStrategy.create(bitmaps);
+                        ((AppCompatActivity) imageView.get().getContext()).runOnUiThread(() -> {
```
Если ты используешь Rx то можно же стандартными шедулерами это делать, тогда и отписка будет на них распространяться

kotiknorm replied:
По моему кстати у любого View есть метод post на главный поток

bobrusha replied:
Добавила AndroidSchedulers.MainThread ✅

kotiknorm commented on app/src/main/java/ru/yandex/yamblz/loader/StubCollageLoader.java:
```diff
@@ -17,9 +33,35 @@ public void loadCollage(List<String> urls, ImageTarget imageTarget) {
     @Override
-    public void loadCollage(List<String> urls, ImageView imageView,
-                            CollageStrategy collageStrategy) {
+    public Subscription loadCollage(List<String> urls, WeakReference<ImageView> imageView,
```
Если в жанрах больше чем 4 исполнителя, то зачем их все грузить?

bobrusha replied:
✅

kotiknorm commented on app/src/main/java/ru/yandex/yamblz/loader/StubCollageLoader.java:
```diff
@@ -17,9 +33,35 @@ public void loadCollage(List<String> urls, ImageTarget imageTarget) {
+
+        return Observable.from(urls)
+                .observeOn(Schedulers.io())
+                .map(this::loadBitmapFromUrl)
```
Тут я понимаю они будут все-таки друг за другом грузиться, хотелось бы параллельно.

bobrusha replied:
✅

kotiknorm commented on app/src/main/java/ru/yandex/yamblz/loader/StubCollageLoader.java:
```diff
@@ -17,9 +33,35 @@ public void loadCollage(List<String> urls, ImageTarget imageTarget) {
+                .subscribe(new Subscriber<LinkedList<Bitmap>>() {
+                    @Override
+                    public void onCompleted() {
+                        Log.d(DEBUG_TAG, "Bitmap size + " + bitmaps.size());
```
Это все по-моему должно быть в onNext

bobrusha replied:
✅

bobrusha committed:
Rewrited loadCollage using advices.

Added using critical section.

Implemented loadCollage(List<String> urls, ImageTarget imageTarget, CollageStrategy collageStrategy)

kotiknorm commented:
Нельзя подписку класть в imageView.setTag?

bobrusha commented:
@metrologma да, в этом случае, конечно, будет void loadCollage, но  тогда всё равно придётся добавить метод setSubscriotion в интерфейс ImageTarget. Если так будет лучше, то можно

bobrusha committed:
Added closing connection to loadBitmapFromUrl()

bobrusha committed:
Fixed bug with critical section.

bobrusha commented:
Возможно, бага происходит потому, что не удаляются ненужные таски.

bobrusha committed:
Fixed bug what replaced birmpas in imageView.
